pkg/dockerregistry/server/auditblobstore.go

```diff
@@ -8,86 +8,179 @@ import (
-	audit.GetLogger(ctx).Log("BlobStore.Put")
+	defer metrics.NewTimer(metrics.RegistryAPIRequests, []string{"blobstore.put", b.repo.Named().Name()}).Stop()
+
+	if audit.LoggerExists(ctx) {
+		audit.GetLogger(ctx).Log("BlobStore.Put")
+	}
+
 	desc, err := b.store.Put(ctx, mediaType, p)
-	audit.GetLogger(ctx).LogResult(err, "BlobStore.Put")
+
+	if audit.LoggerExists(ctx) {
+		audit.GetLogger(ctx).LogResult(err, "BlobStore.Put")
+	}
 	return desc, err
 }
 
 func (b *auditBlobStore) Create(ctx context.Context, options ...distribution.BlobCreateOption) (distribution.BlobWriter, error) {
-	audit.GetLogger(ctx).Log("BlobStore.Create")
+	defer metrics.NewTimer(metrics.RegistryAPIRequests, []string{"blobstore.create", b.repo.Named().Name()}).Stop()
+
+	if audit.LoggerExists(ctx) {
+		audit.GetLogger(ctx).Log("BlobStore.Create")
+	}
+
 	writer, err := b.store.Create(ctx, options...)
-	audit.GetLogger(ctx).LogResult(err, "BlobStore.Create")
-	return &blobWriter{BlobWriter: writer}, err
+
+	if audit.LoggerExists(ctx) {
+		audit.GetLogger(ctx).LogResult(err, "BlobStore.Create")
+	}
+
+	return &blobWriter{
+		BlobWriter: writer,
+		repo:       b.repo,
+	}, err
 }
 
 func (b *auditBlobStore) Resume(ctx context.Context, id string) (distribution.BlobWriter, error) {
-	audit.GetLogger(ctx).Log("BlobStore.Resume")
+	defer metrics.NewTimer(metrics.RegistryAPIRequests, []string{"blobstore.resume", b.repo.Named().Name()}).Stop()
+
+	if audit.LoggerExists(ctx) {
+		audit.GetLogger(ctx).Log("BlobStore.Resume")
+	}
+
 	writer, err := b.store.Resume(ctx, id)
-	audit.GetLogger(ctx).LogResult(err, "BlobStore.Resume")
+
+	if audit.LoggerExists(ctx) {
+		audit.GetLogger(ctx).LogResult(err, "BlobStore.Resume")
+	}
+
 	return writer, err
 }
 
 func (b *auditBlobStore) ServeBlob(ctx context.Context, w http.ResponseWriter, req *http.Request, dgst digest.Digest) error {
-	audit.GetLogger(ctx).Log("BlobStore.ServeBlob")
+	defer metrics.NewTimer(metrics.RegistryAPIRequests, []string{"blobstore.serveblob", b.repo.Named().Name()}).Stop()
+
+	if audit.LoggerExists(ctx) {
+		audit.GetLogger(ctx).Log("BlobStore.ServeBlob")
+	}
+
 	err := b.store.ServeBlob(ctx, w, req, dgst)
-	audit.GetLogger(ctx).LogResult(err, "BlobStore.ServeBlob")
+
+	if audit.LoggerExists(ctx) {
+		audit.GetLogger(ctx).LogResult(err, "BlobStore.ServeBlob")
+	}
+
 	return err
 }
 
 func (b *auditBlobStore) Delete(ctx context.Context, dgst digest.Digest) error {
-	audit.GetLogger(ctx).Log("BlobStore.Delete")
+	defer metrics.NewTimer(metrics.RegistryAPIRequests, []string{"blobstore.delete", b.repo.Named().Name()}).Stop()
+
+	if audit.LoggerExists(ctx) {
+		audit.GetLogger(ctx).Log("BlobStore.Delete")
+	}
+
 	err := b.store.Delete(ctx, dgst)
-	audit.GetLogger(ctx).LogResult(err, "BlobStore.Delete")
+
+	if audit.LoggerExists(ctx) {
+		audit.GetLogger(ctx).LogResult(err, "BlobStore.Delete")
+	}
+
 	return err
 }
 
 type blobWriter struct {
 	distribution.BlobWriter
+
+	repo *repository
```

Comment:
Aren't we avoiding passing the repository everywhere? This could be just the `Named`.